services/kafka/hash.go

```diff
@@ -0,0 +1,116 @@
+package kafka
+
+import (
+	"errors"
+	"hash/crc32"
+	"math/rand"
+	"time"
+
+	"github.com/Shopify/sarama"
+)
+
+var ErrNonPositivePartitions = errors.New("number of partitions must be positive")
+
+type murmur2 struct{}
+
+// Partition here is based on the murmur2 balance function in github.com/segmentio/kafka-go to not break compatibility
+// with earlier versions.
+func (b murmur2) Partition(msg *sarama.ProducerMessage, numPartitions int32) (int32, error) {
+	if numPartitions == 0 {
+		return 0, ErrNonPositivePartitions
+	}
+
+	// NOTE: the murmur2 balancers in java and librdkafka treat a nil key as
+	//       non-existent while treating an empty slice as a defined value.
+	if msg.Key == nil {
+		return rand.New(rand.NewSource(time.Now().UTC().UnixNano())).Int31n(numPartitions), nil
```

Comment:
Building a new source every time seems excessive - should we construct it once and keep it in the murmur2 struct maybe?